merge.py

```diff
@@ -1,29 +1,81 @@
-from typing import Callable
+from functools import partial
+from typing import Any, Callable, Optional, Sequence, Tuple
 
 
-def merge(dict1: dict, dict2: dict, array_strategy: Callable = lambda x, y: x) -> dict:
-    output = {}
+def merge(
+    dict1: dict, 
+    dict2: dict, 
+    dict_strategy: Callable = lambda x, y: x,
+    list_strategy: Callable = lambda x, y: x
+    ) -> dict:
+    """
+    `merge` is a function to recursively, uhh, merge two dictionaries
 
-    # adds keys from `dict2` if they do not exist in `dict1`
-    intersection = {**dict2, **dict1}
+    By default, `merge` operates like the expression ``{**dict1, **dict2}``, 
+    but when supplied functions for `dict_strategy` and `list_strategy` it 
+    is capable of merging any Python dict in any way imaginable. You can 
+    choose from some predefined functions from the `strategies` module to 
+    get started quickly, or roll your own for even more fun. 
 
-    for k_intersect, v_intersect in intersection.items():
-        if k_intersect not in dict1:
-            v_dict2 = dict2[k_intersect]
-            output[k_intersect] = v_dict2
+    TODO: add `strategies` module
+    """
 
-        elif k_intersect not in dict2:
-            output[k_intersect] = v_intersect
+    keys = get_keys(dict1, dict2)
+    return {
+        key: merge_map(
+            *get_values(dict1, dict2, key), 
+            dict_strategy, 
+            list_strategy)
+        for key in keys
+    }
 
-        elif isinstance(v_intersect, dict):
-            v_dict2 = dict2[k_intersect]
-            output[k_intersect] = merge(v_intersect, v_dict2)
 
-        elif isinstance(v_intersect, list):
-            v_dict2 = dict2[k_intersect]
-            output[k_intersect] = array_strategy(v_intersect, v_dict2)
+def get_keys(left: dict, right: dict) -> set:
+    return left.keys() | right.keys()
```

Comment:
That's some cool syntax! Assuming it's a concise way to return all of the keys of both dicts?

Reply:
yep! it's a set union: https://docs.python.org/3.8/library/stdtypes.html#frozenset.union